### PR TITLE
feat!: fine tune user interface

### PR DESCRIPTION
**Pull request by xiaoshihou514**
This pr fine tunes the user interface according to [nvim-best-practices](https://github.com/nvim-neorocks/nvim-best-practices)
- [x] use `Guard` + subcommands instead of multiple individual commands
- [x] remove / deprecate the setup function, `ft(...):fmt` should work immediately
- [ ] ~finer typing for `ft(...):fmt`, `ft(...):lint`, etc.~ HARD, maybe later
- [x] versioning with [luarocks-tag-release](https://github.com/nvim-neorocks/luarocks-tag-release)
- [x] remove some of the `vim.validate` calls for perf
- [x] issue templates from `haskell-tools.nvim`
- [x] allow formatter to be a function that returns the table
- [x] print stderr on fail (reason?)

### CODE REVIEW

**Comment by xiaoshihou514** [comment]
@glepnir I think you would have to add some repo secrets [as instructed here](https://www.github.com/nvim-neorocks/sample-luarocks-plugin)

**Comment by xiaoshihou514** [comment]
It seems like `vim.fn.executable` may return 1 for tools that are installed by mason but not in PATH, I added a check for this in `lua/guard/health.lua`.

File: spec/autocmd_spec.lua [hunk 1]
```diff
@@ -1,12 +1,16 @@
 ---@diagnostic disable: undefined-field, undefined-global
 local api = vim.api
 local ft = require('guard.filetype')
+vim.g.guard_config = {
```
Comment (glepnir) [inline spec/autocmd_spec.lua:4]: I am thingking about a function wrapper like viml get 

```lua
local function opt_get(name)
     return vim.F.if_nil(vim.tbl_get(vim.g.guard_config, name), name == 'lsp_as...' and false or true))
end
```

Comment (xiaoshihou514) [reply spec/autocmd_spec.lua:4]: Hmm, is it better than setting the default in plugin?

Comment (xiaoshihou514) [reply spec/autocmd_spec.lua:4]: Hmm, is it better than setting the default in `plugin/`?

Comment (glepnir) [reply spec/autocmd_spec.lua:4]: then we don't store a table in script . in viml plugin alwasy do `let s:var = get(g:xx, 'xx')` so this is similiar.

Comment (xiaoshihou514) [reply spec/autocmd_spec.lua:4]: Right, will change later.

**Comment by glepnir** [comment]
also create a .luarc.json file for luals config.

**Review by glepnir — APPROVED** [review]

**Comment by glepnir** [comment]
nice work

**Comment by glepnir** [comment]
thinking about create a floating window to show the all config of user defined and config for current buffer. like lspinfo we can call `guard info` or something else.

**Comment by xiaoshihou514** [comment]
> thinking about create a floating window to show the all config of user defined and config for current buffer. like lspinfo we can call `guard info` or something else. 

I went for just printing, though that would be nice.

**Comment by glepnir** [comment]
yup. but no rush. can in the subsequent PR

**Comment by xiaoshihou514** [comment]
Added info in floating markdown window :)

File: plugin/guard.lua [hunk 2]
```diff
@@ -44,26 +44,49 @@ local cmds = {
...
+    for _, formatter in ipairs(formatters) do
+      for _, line in ipairs(vim.split(vim.inspect(formatter), '\n')) do
+        table.insert(lines, line)
+      end
+    end
+    table.insert(lines, '```')
+    table.insert(lines, '')
+    table.insert(lines, '- linters:')
+    table.insert(lines, '')
+    table.insert(lines, '```lua')
```
Comment (glepnir) [inline plugin/guard.lua:71]: use list_extend

File: plugin/guard.lua [hunk 3]
```diff
@@ -44,26 +44,49 @@ local cmds = {
...
+    api.nvim_set_option_value('modifiable', false, { buf = 0 })
   end,
 }
 
 api.nvim_create_user_command('Guard', function(opts)
   local f = cmds[opts.args]
-  _ = not f and vim.notify('Invalid subcommand: ' .. opts.args) or f(opts)
+  if f then
+    f(opts)
+  else
```
Comment (glepnir) [inline plugin/guard.lua:87]:
```suggestion
  if f then
    return f(opts)
  end
```

Comment (xiaoshihou514) [reply plugin/guard.lua:87]: `f` doesn't return anything, only side effects

Comment (glepnir) [reply plugin/guard.lua:99]: what is the difference? There is no need for the else instruction. f will still run and the return value will be nil.

Comment (glepnir) [reply plugin/guard.lua:99]: more clean just with temporary expr.. return f and f or vim.notify

Comment (xiaoshihou514) [reply plugin/guard.lua:99]: This will cause `vim.notify` to be evaluated regardless of whether `f` is nil.

**Pull request merged**